jsonrpcservice/__init__.py

```diff
@@ -0,0 +1,259 @@
+
+        self._raw_request = raw_request
+        self._parsed_request = parsed_request
+
+    def parsed_request(self):
+        if self._parsed_request is None:
+            try:
+                self._parsed_request = json.loads(self._raw_request)
+            except (TypeError, ValueError) as e:
+                raise ParseError(message=str(e))
+
+        return self._parsed_request
+
+    def id(self):
+        self._prepare()
+        return self._id
+
+    def method(self):
+        self._prepare()
+        return self._method
+
+    def args(self):
+        self._prepare()
+        return self._args
+
+    def kwargs(self):
+        self._prepare()
+        return self._kwargs
+
+    def is_notification(self):
+        self._prepare()
+        return self._notification
+
+    def _prepare(self):
+        if self._prepared:
+            return
+
+        self._prepared = True
+
+        data = self.parsed_request()
+
+        if not isinstance(data, dict):
+            raise InvalidRequest('Expected an object as request body. Batch requests are not supported.')
+
+        # Remember `id` as the first thing. `id` can be ommited for notifications.
+        self._id = data.get('id')
+
+        if 'jsonrpc' not in data:
+            raise InvalidRequest('Missing "jsonrpc" member in request.')
+
+        if data['jsonrpc'] != "2.0":
+            raise InvalidRequest('Only JSON-RPC version "2.0" is supported.')
+
+        if self._id is None:
+            # Flag as notification only if client specified protocol correctly and thus should know he can't expect
+            # response if id is not set.
+            self._notification = True
+
+        if 'method' not in data:
+            raise InvalidRequest('Missing "method" member in request.')
+
+        if not isinstance(data['method'], str):
+            raise InvalidRequest('"method" of request must be a string.')
+
+        self._method = data['method']
+
+        if 'params' in data:
+            if isinstance(data['params'], list):
+                self._args.extend(data['params'])
+            elif isinstance(data['params'], dict):
+                self._kwargs.update(data['params'])
+            else:
+                raise InvalidRequest('"params" of request must be either object or array if present.')
+
+
+class Service(object):
+    def __init__(self):
+        self._methods = {}
+
+    def method(self, method):
+        if callable(method):
+            self.add_method(method.__name__, method)
+
+            return method
+        else:
+            def wrapper(func):
+                self.add_method(method, func)
+                return func
+
+            return wrapper
+
+    def add_method(self, method_name, func):
+        if method_name in self._methods:
+            raise ValueError('Method "{}" already registered.'.format(method_name))
+
+        self._methods[method_name] = func
+
+    def add_methods(self, methods, prefix=''):
```

Comment:
Please consider renaming the methods param, because practically it can also be a module, which is a bit misleading.

Reply:
But module is considered as collection of methods, so it still stands.

Reply:
AFAIK not only methods.